### ISSUE TITLE
[x86, debug] `test_remove_garbage_group0_members` failed

### ISSUE THREAD

**Issue by piodul** [scylladb/scylladb#16373]
Failed during CI run of https://github.com/scylladb/scylladb/pull/16337 - the changes in the code unrelated to the failure.

Link to the job: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5124/
Logs from the run: [logz.zip](https://github.com/scylladb/scylladb/files/13636441/logz.zip)

```
manager = <test.pylib.manager_client.ManagerClient object at 0x7f9756022210>

    async def test_remove_garbage_group0_members(manager: ManagerClient):
        """
        Verify that failing to leave group 0 or remove a node from group 0 in removenode/decommission
        can be handled by executing removenode (which should clear the 'garbage' group 0 member),
        even though the node is no longer a token ring member. Does not apply to Raft-topology mode.
        """
        # 4 servers, one dead
        cfg = {'enable_user_defined_functions': False,
               'experimental_features': list[str]()}
        servers = [await manager.server_add(config=cfg) for _ in range(4)]
        removed_host_id = await manager.get_host_id(servers[0].server_id)
        await manager.server_stop_gracefully(servers[0].server_id)
    
        logging.info(f'removenode {servers[0]} using {servers[1]}')
        # removenode will fail after removing the server from the token ring,
        # but before removing it from group 0
        await inject_error_one_shot(manager.api, servers[1].ip_addr,
                                    'removenode_fail_before_remove_from_group0')
        try:
            await manager.remove_node(servers[1].server_id, servers[0].server_id)
        except Exception:
            # Note: the exception returned here is only '500 internal server error',
            # need to look in test.py log for the actual message coming from Scylla.
            logging.info(f'expected exception during injection')
    
        # Query the storage_service/host_id endpoint to calculate a list of known token ring members' Host IDs
        # (internally, this endpoint uses token_metadata)
        token_ring_ids = await get_token_ring_host_ids(manager, servers[1])
        logging.info(f'token ring members: {token_ring_ids}')
    
        group0_members = await get_current_group0_config(manager, servers[1])
        logging.info(f'group 0 members: {group0_members}')
        group0_ids = {m[0] for m in group0_members}
    
        # Token ring members should currently be a subset of group 0 members
        assert token_ring_ids <= group0_ids
    
        garbage_members = group0_ids - token_ring_ids
        logging.info(f'garbage members: {garbage_members}')
        assert len(garbage_members) == 1
        garbage_member = next(iter(garbage_members))
    
        # The garbage member is the one that we failed to remove
        assert garbage_member == removed_host_id
    
        # Verify that at least it's a non-voter.
        assert garbage_member in {m[0] for m in group0_members if not m[1]}
    
        logging.info(f'removenode {servers[0]} using {servers[1]} again')
        # Retry removenode. It should skip the token ring removal step and remove the server from group 0.
        await manager.remove_node(servers[1].server_id, servers[0].server_id)
    
        group0_members = await get_current_group0_config(manager, servers[1])
        logging.info(f'group 0 members: {group0_members}')
        group0_ids = {m[0] for m in group0_members}
    
        # Token ring members and group 0 members should now be the same.
        assert token_ring_ids == group0_ids
    
        # Verify that availability is not reduced.
        # Stop one of the 3 remaining servers and try to remove it. It should succeed with only 2 servers.
    
        logging.info(f'stop {servers[1]}')
        await manager.server_stop_gracefully(servers[1].server_id)
    
        logging.debug(f'waiting for {servers[2]} to see {servers[1]} is down')
        await manager.server_not_sees_other_server(servers[2].ip_addr, servers[1].ip_addr)
        logging.info(f'removenode {servers[1]} using {servers[2]}')
        await manager.remove_node(servers[2].server_id, servers[1].server_id)
    
        # Perform a similar scenario with decommission. One of the node fails to decommission fully,
        # but it manages to leave the token ring. We observe the leftovers using the same APIs as above
        # and remove the leftovers.
        # We can do this with only 2 nodes because during decommission we become a non-voter before
        # leaving the token ring, thus the remaining single node will become a voting majority
        # and will be able to perform removenode alone.
    
        decommissioned_host_id = await manager.get_host_id(servers[2].server_id)
        await manager.api.enable_injection(
            servers[2].ip_addr, 'decommission_fail_before_leave_group0', one_shot=True)
        logging.info(f'decommission {servers[2]}')
        try:
            await manager.decommission_node(servers[2].server_id)
        except Exception:
            logging.info(f'expected exception during injection')
        logging.info(f'stop {servers[2]}')
        await manager.server_stop_gracefully(servers[2].server_id)
    
        token_ring_ids = await get_token_ring_host_ids(manager, servers[3])
        logging.info(f'token ring members: {token_ring_ids}')
    
>       group0_members = await get_current_group0_config(manager, servers[3])

test/topology_custom/test_topology_remove_garbage_group0.py:111: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/topology/util.py:54: in get_current_group0_config
    host = (await wait_for_cql_and_get_hosts(manager.cql, [srv], time.time() + 60))[0]
test/pylib/util.py:77: in wait_for_cql_and_get_hosts
    hosts = await wait_for(get_hosts, deadline)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pred = <function wait_for_cql_and_get_hosts.<locals>.get_hosts at 0x7f97546254e0>
deadline = 1701981325.9481356, period = 1

    async def wait_for(
            pred: Callable[[], Awaitable[Optional[T]]],
            deadline: float, period: float = 1) -> T:
        while True:
>           assert(time.time() < deadline), "Deadline exceeded, failing test."
E           AssertionError: Deadline exceeded, failing test.

test/pylib/util.py:45: AssertionError
```

There is already a similar issue https://github.com/scylladb/scylladb/issues/16306 but the error looks a bit different. Not sure whether it's the same, so I opened a new issue.

**Comment by bhalevy** [comment]
Seen in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7157/artifact/testlog/x86_64/debug/topology_custom.test_topology_remove_garbage_group0.2.log

**Comment by kbr-scylla** [comment]
> Seen in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7157/artifact/testlog/x86_64/debug/topology_custom.test_topology_remove_garbage_group0.2.log

[scylla-968.log](https://github.com/scylladb/scylladb/files/14513518/scylla-968.log)
[scylla-974.log](https://github.com/scylladb/scylladb/files/14513519/scylla-974.log)
[scylla-985.log](https://github.com/scylladb/scylladb/files/14513520/scylla-985.log)
[scylla-993.log](https://github.com/scylladb/scylladb/files/14513521/scylla-993.log)
[topology_custom.test_topology_remove_garbage_group0.2.log](https://github.com/scylladb/scylladb/files/14513523/topology_custom.test_topology_remove_garbage_group0.2.log)

**Comment by tchaikov** [comment]
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7173/artifact/testlog/x86_64/debug/topology_custom.test_topology_remove_garbage_group0.1.log
[scylla-1041.log](https://github.com/scylladb/scylladb/files/14518196/scylla-1041.log)
[scylla-1018.log](https://github.com/scylladb/scylladb/files/14518197/scylla-1018.log)
[scylla-1032.log](https://github.com/scylladb/scylladb/files/14518198/scylla-1032.log)
[topology_custom.test_topology_remove_garbage_group0.1.log](https://github.com/scylladb/scylladb/files/14518199/topology_custom.test_topology_remove_garbage_group0.1.log)

**Comment by kbr-scylla** [comment]
I think I see what's happening

The test is booting nodes, and then immediately starts shutting down nodes and removing them from the cluster.
The shutting down and removing may happen before driver manages to connect to all nodes in the cluster. In particular, the driver didn't yet connect to the last bootstrapped node.
Or it can even happened that the driver has connected, but the control connection is established to the first node, and the driver fetched topology from the first node when the first node didn't yet consider the last node to be normal. So the driver decides to close connection to the last node:
```
22:34:03.159 DEBUG> [control connection] Removing host not found in peers metadata: <Host: 127.42.90.14:9042 datacenter1>
```

Eventually, at the end of the test, only the last node remains, all other nodes have been removed or stopped. But the driver does not have a connection to that last node.

I need to modify the test so that driver has time to establish connections to all nodes at the beginning, before we start shutting down and removing nodes.
